Metrics: quick start docs for Prometheus/Grafana (#2795)
open-telemetry/opentelemetry-dotnet: docs/metrics/getting-started-prometheus-grafana/README.md

# Quick start on exporting metrics to Prometheus/Grafana

- [Quick start on exporting metrics to Prometheus/Grafana](#quick-start-on-exporting-metrics-to-prometheusgrafana)
  - [Prerequisite](#prerequisite)
  - [Introduction](#introduction)
    - [Configure OpenTelemetry to Expose metrics via Prometheus Endpoint](#configure-opentelemetry-to-expose-metrics-via-prometheus-endpoint)
    - [Check Results in the browser](#check-results-in-the-browser)
  - [Download Prometheus](#download-prometheus)
  - [Prometheus and Grafana](#prometheus-and-grafana)
    - [Configuration](#configuration)
    - [Start Prometheus](#start-prometheus)
    - [View Results in Prometheus](#view-results-in-prometheus)
    - [View/Query Results with Grafana](#viewquery-results-with-grafana)

## Prerequisite

It is highly recommended to go over the [getting-started](../getting-started/README.md)

> project under the document folder of metrics before following along this document.

Reviewer: ```suggestion
doc before following along this document.
```

## Introduction

- [What is Prometheus?](https://prometheus.io/docs/introduction/overview/)

- [Grafana support for
  Prometheus](https://prometheus.io/docs/visualization/grafana/#creating-a-prometheus-graph)

### Configure OpenTelemetry to Expose metrics via Prometheus Endpoint

Create a new console application and run it:

```sh
dotnet new console --output prometheus-http-server
cd prometheus-http-server
dotnet run
```

> We will have to add a reference to prometheus exporter to the

Reviewer: ```suggestion
Add a reference to the latest prometheus exporter to the
```

Reviewer: add link to https://www.nuget.org/packages/opentelemetry.exporter.prometheus

prometheus-http-server application.

```shell
dotnet add package OpenTelemetry.Exporter.Prometheus --version 1.2.0-rc1
```

Now, we are going to make some small tweaks to the example in the
getting-started metrics `Program.cs` to make the metrics available via
OpenTelemetry Prometheus Exporter.

First, copy and paste everything from getting-started

> metrics[example](../getting-started/Program.cs) to the Program.cs file of the

Reviewer 2: Add a space between "metrics" and "[example]".

export data via the endpoint defined by `HttpListenerPrefixes`.

Also, for our learning purpose, use a while-loop to keep increasing the counter
value until any key is pressed.

```csharp
Console.WriteLine("Press any key to exit");
while (!Console.KeyAvailable)
{
    Thread.Sleep(1000);
    MyFruitCounter.Add(1, new("name", "apple"), new("color", "red"));
    MyFruitCounter.Add(2, new("name", "lemon"), new("color", "yellow"));
    MyFruitCounter.Add(1, new("name", "lemon"), new("color", "yellow"));
    ...
    ...
    ...
}
```

After the above modifications, now our `Program.cs` should look like [this](./Program.cs).

### Check Results in the browser

Start the application and leave the process running. Now we should be able to
see the metrics at the endpoint we've defined in `Program.cs`; in this case, the
endpoint is: "http://localhost:9184/".

Check the output metrics with your favorite browser:

![MyFruitCounter output:](https://user-images.githubusercontent.com/16979322/150242010-8bde0002-44a5-4c84-94e6-3e0ee8a6ea4f.PNG)

Now, we understand how we can configure Opentelemetry `PrometheusExporter` to

> export metrics the endpoint we sepcified. Next, we are going to learn about how

Reviewer: ```suggestion
export metrics the endpoint we specified. Next, we are going to learn about how
```

> to use Prometheus and Grafana to view/query our metrics with powerful

Reviewer: ```suggestion
to use Prometheus and Grafana to view/query the metrics
```

    - localhost:9184
```

### Start Prometheus

Follow the instructions from
[starting-prometheus](https://prometheus.io/docs/introduction/first_steps/#starting-prometheus)
to start the Prometheus server and verify it has been started successfully.

Please note that we will need pass in otel.yml file as the argument:

```console
./prometheus --config.file=otel.yml
```

### View Results in Prometheus

To use the graphical interface for viewing our metrics with Prometheus, navigate
to "http://localhost:9090/graph", and type `MyFruitCounter` in the expression
bar of the UI; finally, click the execute button.

We should be able to see the following chart from the browser:

![Prometheus Graph:](https://user-images.githubusercontent.com/16979322/150242083-65b84f25-c95f-4e9b-a64f-699ad8816602.PNG)

From the legend, we can see that the `instance` name and the `job` name are the
values we have set in `otel.yml`.

Congratulations!

Now we know how to configure Prometheus server and deploy OpenTelemetry
`PrometheusExporter` to export our metrics. Next, we are going to explore a tool
called Grafana, which has powerful visualizations for the metrics.

### View/Query Results with Grafana

> First of all, please [Install Grafana](https://grafana.com/docs/grafana/latest/installation/).

Reviewer 2: Do we want to add "First of all," here? 😂

Author: Not really :innocent:, removed.
